src/get-static-value.mjs

```diff
@@ -247,6 +247,29 @@ function getElementValues(nodeList, initialScope) {
     return valueList
 }
 
+/**
+ * Returns whether the given variable definition is either `const` or new written to.
+ * @param {import("eslint").Scope.Variable} variable
+ * @param {import("eslint").Scope.Definition & { type: "Variable" }} def
+ * @returns {boolean}
+ */
+function isEffectivelyConst(variable, def) {
```

Comment:
I think it would be better to omit the `def` argument and check in this function (by moving the check done in L427) whether the number of definitions is one.
This is because I think it is easier to understand if multiple `var` declarations are checked by this function.
What do you think?

Reply:
I'm not entirely sure what you mean, but from what I understand, you want to make `isEffectivelyConst` entirely about the variable and not about any specific definition of it. If so, then I agree. Good idea.

Reply:
Sorry, I didn't explain it well.
I intended code like this:

```js
function isEffectivelyConst(variable) {
    if (variable.defs.length !== 1) {
        return false
    }
    const def = variable.defs[0]
    if (def.type !== "Variable") {
        return false
    }
    if (
        def.parent &&
        def.parent.kind === "const"
    ) {
        return true
    }
    const refs = variable.references

    const inits = refs.filter((r) => r.init).length
    const reads = refs.filter((r) => r.isReadOnly()).length
    if (inits === 1 && reads + inits === refs.length) {
        // there is only one init and all other references only read
        return true
    }
    return false
}

// ...

// Constants.
if (variable != null && isEffectivelyConst(variable)) {
    const def = variable.defs[0]
    // TODO(mysticatea): don't support destructuring here.
    if (def.node.id.type === "Identifier") {
        return getStaticValueR(def.node.init, initialScope)
    }
}
```

Note that the actual behavior has not changed between the current code and the changed code. This is a suggestion for improving readability.
The change allows isEffectivelyConst to independently check whether a variable is constant. So it's easier for me to read.
What do you think?

Reply:
I don't think that this is a step in the right direction. My main issue is that the purpose of `isEffectivelyConst` is not clear anymore. Previously, `isEffectivelyConst` checked one thing: whether a variable (or variable definition) is effectively const. 

But what does your version do? The behavior of `isEffectivelyConst` got a lot more complex because it checks const-ness (like before), and it also checks the number of definitions and their shape. 
From what I see, you merged 2 function into one. One that checks that a variable only has one definition of a certain type (similar to our [`findSimpleVariable`](https://github.com/ota-meshi/eslint-plugin-regexp/blob/8755332e6ec5d03d6c8f48ebc00b7f5c09b8e662/lib/utils/ast-utils/utils.ts#L54)), and one that checks for const-ness.

Reply:
What I found `isEffectivelyConst` to be a bit confusing was that I needed to assume that the variable definition was one before using `isEffectivelyConst`.

For example if you have a TS like:

```ts
let foo =  42;
let foo;

console.log(foo)
```

(It's a parse error with common JS parsers like `espree`, but it's a valid JS with the `typescript-eslint/parser`.)

The current `isEffectivelyConst` returns true for the `foo` variable, but I'm not sure if it's effectively const.
So I thought it would be better if `isEffectivelyConst` checks the number of variables defined.
However, I agree with your concerns about `isEffectivelyConst` becoming more complicated.